docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

```diff
@@ -0,0 +1,1937 @@
+  a previous version of PQS, or add the following filter predicate to your query:
+  ``SELECT \* FROM active('my_package:My.App:MyTemplate') WHERE package_id = 'my_package_id'``
+
+Given that PQS uses a document-oriented model for ledger content
+(JSONB), extensions to contract payloads are handled simply by returning
+the additional data in the blob.
+
+Daml Shell
+----------
+
+Daml Shell builds on PQS by providing a shell interface to inspect the
+ledger using package-name to view all versions of contracts, in an
+integrated way.
+
+Daml-Script-Beta
+----------------
+
+Daml 2.9 introduces a new version of Daml Script, which can be used by
+depending on ``daml-script-beta`` in your ``daml.yaml``, as you will have seen
+in `Writing your first upgrade <#writing-your-first-upgrade>`__. Only this version of Daml Script
+supports upgrades over the Ledger API.
+
+All commands and queries in this version of Daml Script now use
+upgrades/downgrades automatically, to ensure that the correct versions
+of choices are exercises, and correct payloads are returned.
+
+The following additional functionality is available for more advanced
+uses of SCU.
+
+**Exact commands**
+
+Each of the four submission commands now has an “exact” variant, of the
+forms ``createExactCmd``, ``exerciseExactCmd``, ``exerciseByKeyExactCmd`` and
+``createAndExerciseExactCmd``. These commands force the participant to
+use the exact version of the package that your script uses, this is most
+useful when you want to be absolutely certain of the choice code you are
+calling. Note that exact and non-exact commands can be mixed in the same
+submission.
+
+Daml Studio support
+-------------------
+
+Daml Studio runs a reference model of Canton called the IDE Ledger, this
+ledger has been updated to support most of the functionality of upgrades
+in daml-script. The behavior that this ledger does not implement is the
+following:
+
+-  Upgrades type checking
+   The Upgrades type errors you get when running daml build are not
+   currently shown as code intelligence in Daml Studio.
+
+-  Contract ID verification
+   This may allow some contract metadata changes to take effect that
+   Canton otherwise would not.
+
+-  Per-submission package preference
+   Any submission to the IDE Ledger mimics a Canton participants
+   default package preference of the most recent package version.
+
+Testing
+=======
+
+Upgrade validity checking
+-------------------------
+
+While the compile time upgrade checks are quite thorough, they are not
+yet a perfect mirror of the checks that a participant does when a
+package is uploaded. We recommend that as a final check for the validity
+of your upgraded package, you can either:
+
+-  Run a Canton sandbox (running ``daml sandbox``) and upload your old and
+   new packages (``daml ledger upload-dar``).
+
+-  Perform a dry-run upload of your package to a more permanent testing
```

Comment:
@basvangijzel-DA @paulbrauner-da : note that there are often multiple packages that are used by a single app, and that need to be changed for an upgrade. See this example of an upgrade to the splice apps: https://github.com/DACH-NY/canton-network-node/commit/917a631b19a264f3a98957d9c9018e3ee9ae8d62

From what I see, the dry-run functionality does not support checking the validity in that case, as it only accepts a single argument. Is that correct?

Reply:
This is true, it would be possible to test all your packages together if you bundled them into one dar, but the UX wouldn't be great. We may want to reconsider this advice.

Reply:
Thanks for confirming. I'll advise to use a fresh participant instead and uploading all packages there in the upgrading guidelines here: https://docs.google.com/document/d/10ttHyhufCI7hGEsd0CA2VCjsXPhAg7Hg-fVeMQOKS4U/edit#bookmark=id.j6ol6es6xln

Reply:
That said, if your several packages interact with eachother, they'd be dependencies, so the upgrade check would run. if you have several disconnected packages, then the upgrade checks don't need to know about them all. This might be fine, @dylant-da can you confirm?

Reply:
Yeah, that might work. It still requires uploading all maximum packages in the dependency ordering; and that needs to be updated in case that hierarchy changes. I'd thus suggest to just strike this suggestion and advise for the one with the sandbox which will always work, and is very simple to use.

Reply:
Also note that the one with a Canton sandbox comes with the caveat that it might use a different version of the upload checks than the one on the participant node that's actually used in production. So ideally we recomment starting a fresh instance of a participant node of the same version that's used in production; and hint at the sandbox as simple to use approximation to that.

Reply:
Yes, if the packages are disconnected then the upgrade checks won't need to know about them all in unison.

Reply:
This recommendation should make it clear that this function is only an alternative to the compiler checks, and then explain that for a multi-DAR workflow a full participant upload is recommended.

Something like:

> We recommend that as a further check for the validity
of your upgraded package, you perform a dry-run upload of your package to a testing environment, using the ``--dry-run`` flag of the ``daml ledger upload-dar`` command, which runs the upgrade type-checking, but does not persist your package to the ledger.

> For workflows involving multiple DARs, we recommend more robust testing by running a Canton sandbox with the same version and environment as your in-production participant and uploading all the old and new packages that constitute your Daml app.

Reply:
@dylant-da : I was under the impression that the full upload is always required, as it does check more than what the compiler checks. Isn't that the case?

Reply:
Yeah I think we should always recommend doing an upload, but the dry-run flag should be useful for uploading to a test environment.
The upload checks do currently check more than the compiler checks, but a dry-run upload runs the same checks as without dry-run.

Reply:
The word "alternative" was poorly chosen - I meant "extension" - the participant checks are more comprehensive than the compiler checks. The dry-run flag always runs the same checks as a full upload would create.

Reply:
I have updated the section with Dylan's suggestion. If we want further changes, lets put them in a different PR, I'd like to merge this PR today.

Reply:
OK. Seems we're mostly aligned then. Let's see whether further changes are required once this PR is merged.